tests/python/test_bit_operations.py

```diff
@@ -29,3 +29,19 @@ def sar(a: ti.i32, b: ti.i32) -> ti.i32:
     # for negative number
     for i in range(n):
         assert sar(neg_test_num, i) == -2**(n - i)
+
+
+@ti.test()
+def test_bit_shr():
+    @ti.kernel
+    def shr(a: ti.i32, b: ti.i32) -> ti.i32:
+        return ti.bit_shr(a, b)
+
+    n = 8
+    test_num = 2**n
+    neg_test_num = -test_num
+    for i in range(n):
+        assert shr(test_num, i) == 2**(n - i)
+    for i in range(n):
+        offset = 0x100000000 if i > 0 else 0
+        assert shr(neg_test_num, i) == (neg_test_num + offset) >> i
```

Comment:
This test will fail on backends other than LLVM due to unimplemented op.

Reply:
This includes(but not limited to...) cc, metal and opengl.

Reply:
> So that backend developers don't need to worry about SAR. This is a very late pass in the compilation process - we still need SAR in the IR for certain domain-specific optimizations.

Good idea! I'll later add uint support to OpenGL so that SHR works.

Reply:
@archibate That's what I'd also like to discuss. I did some experiments, in the LLVM backend, `bit_sar` is implemented using `CreateAShr` which simply copies the MSB without considering the type. While on other backends, `bit_sar` is expressed with `>>` which will behave differently according to unsigned/signed information. I think we need to decide which kind of SAR operation we truly want, either a pure one copying the MSB or a one that is more similar to the `>>` in C.

Reply:
Thanks for all the discussions. My two cents:
 - In the frontend `>>` always translates to `BinaryOpType::bit_sar` and `ti.bit_shr` always translates to `BinaryOpType::bit_shr`.
 - In `demote_operations.cpp` we convert `bit_shr` on signed integers into three sub-operations as discussed above. Then we only have `bit_sar` for the backend.
 - In the backend, since we only have `bit_sar`, its behavior is determined by the type of its operands.

Reply:
@yuanming-hu Thanks for the clarification. The point(or problem) I'd like to discuss is the behavior of `bit_sar` which `bit_shr` will eventually rely on. As I wrote in the above comment, in the LLVM backend, `bit_sar` is implemented with a `CreateAShr` which directly maps to a `sar` instruction and simply copies the MSB, that is, it will ignore the type information so casting signed to unsigned will not work since the low-level bits are not changed by type casts. In other backends (at least Metal), `bit_sar` is implemented using the operator `>>`, which will consider the type information, that is, it will map to a `zext` and `sar` when working on unsigned integers. This is creating different behaviors across backends. We should first confirm what kind of behavior we want for `bit_sar` before we move on.

Reply:
Let's follow the Metal behavior `bit_sar` in backends. And provide `bit_shr` using `demote_operations` as described above.
The LLVM backend should do some branches to match the behavior of Metal, e.g.:
```cpp
// on bit_sar
if (dtype == unsigned) {
  CreateAShr();
} else {
  CreateLShr();
}
```

Reply:
Perhaps one source of confusion is that LLVM doesn't distinguish signed/unsigned integers (so does hardware such x64). So `SHR` always shifts everything and `SAR` copies the MSB in LLVM.

In Taichi, on unsigned integers `bit_shr=bit_sar`. On signed integers `bit_shr` = `SHR` in LLVM and `bit_sar` = `SAR` in LLVM.